plugins/kmeans.cc

```diff
@@ -294,72 +303,79 @@ KmeansFunctionConfigDescription()
 
 
 /*****************************************************************************/
-/* KMEANS FUNCTION                                                              */
+/* KMEANS FUNCTION                                                           */
 /*****************************************************************************/
 
+DEFINE_STRUCTURE_DESCRIPTION(KmeansFunctionArgs);
+
+KmeansFunctionArgsDescription::
+KmeansFunctionArgsDescription()
+{
+    addField("embedding", &KmeansFunctionArgs::embedding,
+             "Embedding values for the k-means function.  The column names in "
+             "this embedding must match those used in the original kmeans "
+             "dataset.");
+}
+
+DEFINE_STRUCTURE_DESCRIPTION(KmeansExpressionValue);
+
+KmeansExpressionValueDescription::KmeansExpressionValueDescription()
+{
+    addField("cluster", &KmeansExpressionValue::cluster,
+             "Index of the row in the `centroids` dataset whose columns describe "
+             "the point which is closest to the input according to the `metric` "
+             "specified in training.");
+}
+
+
 struct KmeansFunction::Impl {
     ML::KMeans kmeans;
-    
-    Impl(const Url & modelFileUrl) {
-        kmeans.load(modelFileUrl.toString());
+    std::vector<ColumnName> columnNames;
+
+    Impl(const Url & modelFileUrl)
+    {
+        filter_istream stream(modelFileUrl.toString());
+        std::string firstLine;
+        std::getline(stream, firstLine);
+        Json::Value md = Json::parse(firstLine);
+        if (md["algorithm"] != "MLDB k-Means model") {
+            throw HttpReturnException(400, "Model file is not a k-means model");
+        }
+        if (md["version"].asInt() != 1) {
+            throw HttpReturnException(400, "k-Means model version is wrong");
```

Comment:
This is very minor but I will at a minimum clarify why I think this is a 500.
My interpretation is that the first error is a 400 (a user provided the wrong model) but the second one is a 500 since, as a user, my expectation is that the latest version of a software should support all previous versions of its artefacts.  Here a user provided the right model but MLDB does not know how to read it.